rs/nns/cmc/src/main.rs

```diff
@@ -1569,6 +1569,62 @@ fn memo_to_intent_str(memo: Memo) -> String {
     }
 }
 
+/// Returns Ok if transaction matches expected_memo.
+///
+/// First, looks at the memo field. If that does not match, falls back to the
+/// icrc1_field. If that is of length 8, converts assuming little endian, and if
+/// that matches, returns Ok.
+fn transaction_has_expected_memo(
+    transaction: &Transaction,
+    expected_memo: Memo,
+) -> Result<(), NotifyError> {
+    fn stringify_memo(memo: Memo) -> String {
+        format!("{} ({})", memo_to_intent_str(memo), memo.0)
+    }
+
+    if transaction.memo == expected_memo {
+        return Ok(());
+    }
+
+    // Fall back to icrc1_memo.
+
+    // Read the field.
+    let Some(icrc1_memo) = &transaction.icrc1_memo else {
+        return Err(NotifyError::InvalidTransaction(format!(
+            "The transaction's memo ({}) does not have the required value ({}).",
```

Comment:
Is the `icrc1_memo` field always required, or only along this code path?

Reply:
It is not guaranteed to be Some, and we do not need it to be. It is just a fallback, in case someone (very reasonably) uses it instead of the legacy memo field.

Ofc, if it is None, we have nothing to fall back to. That is why this error message does not mention icrc1_memo. It just says that the (legacy) memo field indicates a different purpose for the transaction.